<?xml version="1.0"?>

<!-- 2017年11月01日　秋本 -->
<!-- つくばチャレンジ本番で使用するlaunch -->
<launch>

  <!-- base_link to base_footprint  -->
  <node pkg="tf" type="static_transform_publisher" name="basefootprint_to_baselink" args="0 0 0 0 0 0 base_footprint base_link 100"/>


  <!-- map load -->
  <arg name="map_name" value="/home/robo/2018_tsukuba/20181111_park_last/map_rere_clean.yaml"/>
  <node pkg="map_server" type="map_server" name="map" args="$(arg map_name)"/>


  <!-- tf remapper -->
  <node pkg="tf" type="tf_remap" name="tf_remapper" output="screen">
    <rosparam param="mappings">
      [{old: base_footprint_odomgyro, new: base_footprint}]
    </rosparam>
  </node>


  <!-- urg_node_front  -->
  <node pkg="urg_node" type="urg_node" respawn="false" name="urg_node_front" output="screen">
    <param name="serial_port" value="/dev/ttyACM1"/>
    <param name="frame_id" value="laser_front"/>
    <param name="angle_min" value="-1.1345"/>
    <param name="angle_max" value="1.1345"/>
    <remap from="scan" to="scan_front"/>
  </node>
  <node pkg="tf" type="static_transform_publisher" name="laser_to_base_link_front" args="0.10 0.0 0.20 0.0 0.0 3.14  /base_link /laser_front 100" />


  <!-- urg_node_back  -->
  <node pkg="urg_node" type="urg_node" respawn="false" name="urg_node_back" output="screen">
    <param name="serial_port" value="/dev/ttyACM2"/>
    <param name="frame_id" value="laser_back"/>
    <param name="angle_min" value="-2.2689"/>
    <param name="angle_max" value="2.2689"/>
    <remap from="scan" to="scan_back"/>
  </node>
  <node pkg="tf" type="static_transform_publisher" name="laser_to_base_link_back" args="-0.45 0.0 0.30 3.14 0.0 0.0 /base_link /laser_back 100" />


  <!-- laser_merge launchに混ぜれるようにコードを変えました -->
  <node pkg="ira_laser_tools" name="laserscan_multi_merger" type="laserscan_multi_merger" output="screen">
    <param name="destination_frame" value="/base_link"/>
    <param name="scan_destination_topic" value="/scan"/>
    <param name="laserscan_topics" value ="/scan_front /scan_back" />
  </node>


  <!-- webcamera -->
  <node pkg="uvc_camera" type="uvc_camera_node" name="uvc_camera" output="screen">
    <param name="width" type="int" value="640" />
    <param name="height" type="int" value="480" />
    <param name="fps" type="int" value="10" />
    <param name="frame" type="string" value="wide_stereo" />
    <param name="auto_focus" type="bool" value="False" />
    <param name="focus_absolute" type="int" value="0" />
    <param name="device" type="string" value="/dev/video0" />
    <param name="camera_info_url" type="string" value="file://$(find uvc_camera)/example.yaml" />
  </node>


  <!-- declare arguments with default values -->
  <arg name="calibration" default="$(find velodyne_pointcloud)/params/VLP-16-HiRes.yaml"/>
  <arg name="device_ip" default="" />
  <arg name="frame_id" default="velodyne" />
  <arg name="manager" default="$(arg frame_id)_nodelet_manager" />
  <arg name="max_range" default="130.0" />
  <arg name="min_range" default="0.4" />
  <arg name="pcap" default="" />
  <arg name="port" default="2368" />
  <arg name="read_fast" default="false" />
  <arg name="read_once" default="false" />
  <arg name="repeat_delay" default="0.0" />
  <arg name="rpm" default="600.0" />

  <!-- start nodelet manager and driver nodelets -->
  <include file="$(find velodyne_driver)/launch/nodelet_manager.launch">
    <arg name="device_ip" value="$(arg device_ip)"/>
    <arg name="frame_id" value="$(arg frame_id)"/>
    <arg name="manager" value="$(arg manager)" />
    <arg name="model" value="VLP16"/>
    <arg name="pcap" value="$(arg pcap)"/>
    <arg name="port" value="$(arg port)"/>
    <arg name="read_fast" value="$(arg read_fast)"/>
    <arg name="read_once" value="$(arg read_once)"/>
    <arg name="repeat_delay" value="$(arg repeat_delay)"/>
    <arg name="rpm" value="$(arg rpm)"/>
  </include>

  <!-- start cloud nodelet -->
  <include file="$(find velodyne_pointcloud)/launch/cloud_nodelet.launch">
    <arg name="calibration" value="$(arg calibration)"/>
    <arg name="manager" value="$(arg manager)" />
    <arg name="max_range" value="$(arg max_range)"/>
    <arg name="min_range" value="$(arg min_range)"/>
  </include>

  <node pkg="tf" type="static_transform_publisher" name="velodyne_to_base_link" args="0.0 0.0 0.0 0.0 0.0 0.0  /base_link /velodyne 100" />


  <!--okatech-->
  <node pkg="okatech_kuaro" type="okatech_kuaro" name="okatech_kuaro" output="screen">
    <param name="sub_serial_topic" value="/serial_receive"/>
    <param name="sub_cmdvel_topic" value="/cmd_vel"/>
    <param name="pub_odom_imu_topic" value="/odom_default"/>
    <param name="frame_id" value="odom_default"/>
    <param name="child_frame_id" value="base_footprint_odom_default"/>
    <!-- <remap from="tf" to="tf_old"/> -->
  </node>-->


  <!--serial-->
  <node pkg="serial_communication" type="serial_communication" name="serial_communication" output="screen"/>


  <!--imu-->
  <node name="microstrain_3dm_gx5_25_node" pkg="microstrain_3dm_gx5_45" type="microstrain_3dm_gx5_25_node" output="screen">
    <param name="port" value="/dev/ttyACM0" type="str" />
    <param name="baudrate" value="115200" type="int" />
    <param name="device_setup" value="true" type="bool" />
    <!-- General Settings -->
    <param name="readback_settings" value="true" type="bool" />
    <param name="save_settings" value="false" type="bool" />
    <param name="auto_init" value="true" type="bool" />
    <!-- AHRS Settings -->
    <param name="publish_imu" value="false" type="bool" />
    <param name="imu_raw_rate" value="50" type="int" />
    <param name="imu_frame_id" value="imu" type="str" />
    <param name="imu_raw_topic" value="imu_raw/data" type="str" />
    <!-- Setting manual declination is not yet implemented/testes -->
    <!-- Declination source 1=None, 2=magnetic, 3=manual -->
    <param name="declination_source" value="2" type="int" />
    <param name="declination" value="0.23" type="double" />
    <param name="publish_imu_filltered" value="true" type="bool" />
    <param name="imu_filtered_rate" value="500" type="int" />
    <param name="imu_filtered_topic" value="imu_filtered/data" type="str" />
  </node>

  <node pkg="tf" type="static_transform_publisher" name="imu_to_base_link" args="0 0 0 0 3.14159 0 /base_link /imu 100" />


  <!--imu_transform-->
  <node pkg="imu_transformer" type="imu_transformer_node" name="imu_data_transformer" output="screen">
    <remap from="imu_in" to="imu_filtered"/>
    <remap from="imu_out" to="imu_out"/>
    <param name="target_frame" value="base_link"/>
  </node>


  <!-- odom_imu_combine -->
  <node pkg="odom_imu_combine" type="odom_imu_combine" name="odom_imu_combine" output="screen">
    <param name="sub_odom_topic" value="/odom_default"/>
    <param name="sub_imu_topic" value="/imu_out/data"/>
    <param name="pub_odom_imu_topic" value="/odom_gyro"/>
    <param name="frame_id" value="odom_gyro"/>
    <param name="child_frame_id" value="base_footprint_odomgyro"/>
    <remap from="tf" to="tf_old"/>
  </node>


  <!-- emergency stop -->
  <!-- <node pkg="recovery_myself" type="Emergency_stop" name="Emergency_stop" output="screen">
    <param name="scan_topic" value="/scan"/>
    <param name="avort_limit" value="5.0"/>
    <param name="scan_range_limit" value="0.5"/>
    <param name="obstacle_count" value="100"/>
  </node> -->


  <!-- pointcloud_to_laserscan -->
  <node pkg="pointcloud_to_laserscan" type="pointcloud_to_laserscan_node" name="pointcloud_to_laserscan">
    <remap from="cloud_in" to="velodyne_points"/>
    <remap from="scan" to="velodyne_scan"/>
    <rosparam>

      target_frame: base_link
      transform_tolerance: 0.01
      min_height: -0.80
      max_height: 10.00
      angle_min: -3.141592
      angle_max: 3.141592
      range_min: 1.00
      range_max: 100.0
      use_inf: true
      concurrency_level: 1

      <!-- target_frame: base_link
      transform_tolerance: 0.01
      min_height: 0.0
      max_height: 10.00
      angle_min: -3.141592
      angle_max: 3.141592
      range_min: 1.00
      range_max: 100.0
      use_inf: true
      concurrency_level: 1 -->
    </rosparam>
  </node>


  <!-- amcl -->
  <node pkg="amcl" type="amcl" name="amcl">
    <remap from="scan" to="velodyne_scan"/> <!-- scanを何にremapするかによって使用するデータが変わる.topic名を入れる velodyne_scanにすればvelodyneから　scanにすればLRFから-->
    <!-- フィルタ関連パラメータ -->
    <!-- 最小のパーティクル数 -->
    <param name="min_particles" value="500"/>
    <!-- 最大のパーティクル数 (多いほど良い、PCのパフォーマンスに応じて設定 -->
    <param name="max_particles" value="4000"/>
    <!-- 実際の分布と推定された分布の間の最大エラー -->
    <param name="kld_err" value="0.05"/>
    <param name="kld_z" value="0.99"/>
    <!-- フィルタの更新に必要な並進運動 (メートル単位)　-->
    <param name="update_min_d" value="0.2"/>
    <!-- フィルタの更新に必要な回転運動 (ラジアン単位) -->
    <param name="update_min_a" value="0.2"/>
    <!-- 再サンプリング間隔 -->
    <param name="resample_interval" value="1"/>
    <!-- 変換許容時間 (秒単位) -->
    <param name="transform_tolerance" value="0.1"/>
    <!-- 指数の減少率 (slow average weight filter) -->
    <param name="recovery_alpha_slow" value="0.0"/>
    <!-- 指数の減少率 (fast average weight filter) -->
    <param name="recovery_alpha_fast" value="0.0"/>
    <!-- スキャンや移動経路の情報を視覚的に表示する周期 (10Hz = 0.1秒) -->
    <param name="gui_publish_rate" value="1.0"/>
    <!-- 距離センサパラメータ -->
    <!-- レーザーの最大距離 (センサに合わせて設定する、メートル単位) -->
    <param name="laser_max_range" value="100.0"/> <!-- 10??? 2016_09_26 -->
    <!-- フィルタが更新されるときに使用されている最大のレーザービームの数 -->
    <param name= "laser_max_beams" value= "30"/>
    <!-- センサモデルのz_hit混合重み (micture weight) -->
    <param name="laser_z_hit" value="0.5"/>
    <!-- センサのz_short混合重み (micture weight) -->
    <param name="laser_z_short" value="0.05"/>
    <!-- センサのz_max混合重み (micture weight) -->
    <param name="laser_z_max" value="0.05"/>
    <!-- センサのz_rand混合重み (micture weight) -->
    <param name="laser_z_rand" value="0.5"/>
    <!-- センサのz_hitを使用したガウスモデルの標準偏差 -->
    <param name="laser_sigma_hit" value="0.2"/>
    <!-- センサのz_shortの指数関数パラメータ -->
    <param name="laser_lambda_short" value="0.1"/>
    <!-- likelihood_field方式センサのための障害物との最大距離 -->
    <param name="laser_likelihood_max_dist" value="2.0"/>
    <!-- センサタイプ (likelihood_fieldとbeam選択) -->
    <param name="laser_model_type" value="likelihood_field"/>
    <!-- オドメトリ関連パラメータ -->
    <!-- ロボット移動方式の "diff"と "omni"が選択可能である。 -->
    <param name="odom_model_type" value="diff"/>
    <!-- 回転運動のとき、予想されるオドメトリの回転運動量推定ノイズ -->
    <param name="odom_alpha1" value="0.2"/>
    <!-- 並進運動のとき、予想されるオドメトリの回転運動量推定ノイズ -->
    <param name="odom_alpha2" value="0.2"/>
    <!-- 並進運動のとき、予想されるオドメトリの並進運動量推定ノイズ -->
    <param name="odom_alpha3" value="0.8"/> <!-- 0.4 -->
    <!-- 回転運動のとき、予想されるオドメトリの並進運動量推定ノイズ -->
    <param name="odom_alpha4" value="0.2"/>
    <!-- オドメトリフレームID -->
    <param name="odom_frame_id" value="odom_gyro"/>
  </node>


  <!-- planner -->
  <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
    <rosparam file="$(find 2018_tsukuba_challenge)/param_teb_test/move_base_params.yaml" command="load"/>
    <rosparam file="$(find 2018_tsukuba_challenge)/param_teb_test/costmap_common_global_params.yaml" command="load" ns="global_costmap"/>
    <rosparam file="$(find 2018_tsukuba_challenge)/param_teb_test/costmap_common_local_params.yaml" command="load" ns="local_costmap"/>
    <rosparam file="$(find 2018_tsukuba_challenge)/param_teb_test/local_costmap_params.yaml" command="load"/>
    <rosparam file="$(find 2018_tsukuba_challenge)/param_teb_test/global_costmap_params.yaml" command="load"/>
    <rosparam file="$(find 2018_tsukuba_challenge)/param_teb_test/base_local_planner_params.yaml" command="load"/>
    <rosparam file="$(find 2018_tsukuba_challenge)/param_teb_test/base_global_planner_params.yaml" command="load"/>
    <rosparam file="$(find 2018_tsukuba_challenge)/param_teb_test/move_slow_and_clear_params.yaml" command="load"/>
    <rosparam file="$(find 2018_tsukuba_challenge)/param_teb_test/recovery_mode_params.yaml" command="load"/>
  </node>


  <!-- rosbag -->
  <node pkg="rosbag" type="record" name="test_record" args="--lz4 -O /home/robo/2018_tsukuba/20181111_park_last/data_out.bag /scan /scan_front /scan_back /velodyne_scan /velodyne_points /tf /imu_filtered/data /imu_out/data /odom_gyro /odom_default /move_base/TebLocalPlannerROS/global_plan /move_base/TebLocalPlannerROS/local_plan /cmd_vel /image_raw" output="screen"/>


  <!-- rviz -->
  <node pkg="rviz" type="rviz" name="rviz" args="-d /home/robo/catkin_ws/src/2018_tsukuba_challenge/rviz/move_base_2018.rviz"/>

</launch>
